group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -3500,6 +3503,56 @@ public void maybeDeleteGroup(String groupId, List<Record> records) {
         }
     }
 
+    /**
+     * A group can be upgraded offline if it's a classic group and empty.
+     *
+     * @param groupId The group to be validated.
+     * @return true if the offline upgrade is valid.
+     */
+    private boolean validateOfflineUpgrade(String groupId) {
+        Group group = groups.get(groupId);
+        return group != null && group.type() == CLASSIC && group.isEmpty();
+    }
+
+    /**
+     * A group can be downgraded offline if it's a consumer group and empty.
+     *
+     * @param groupId The group to be validated.
+     * @return true if the offline downgrade is valid.
+     */
+    private boolean validateOfflineDowngrade(String groupId) {
+        Group group = groups.get(groupId);
+        return group != null && group.type() == CONSUMER && group.isEmpty();
+    }
+
+    /**
+     * Upgrade the empty group if it's valid.
+     *
+     * @param groupId The group id to be migrated.
+     * @param records The list of records to delete the previous group.
+     */
+    public void maybeUpgradeEmptyGroup(String groupId, List<Record> records) {
+        if (validateOfflineUpgrade(groupId)) {
+            deleteGroup(groupId, records);
+            removeGroup(groupId);
```

Comment:
nit: Those two methods next to each others look a bit weird. I wonder if we should rename `deleteGroup` to `createGroupTombstoneRecords`.

Another thing is that we actually get the group from the map in each method called here. I wonder if we should inline and simplify it. For instance, we could lookup the group as the first thing in this method, check the condition, then call `createGroupTombstoneRecords` on the group and finally remove the group from the map. What do you think?